pkl-core/src/main/java/org/pkl/core/ast/expression/literal/AmendFunctionNode.java

```diff
@@ -76,8 +79,17 @@ public AmendFunctionNode(
       //
       // As a remedy, we simply copy outer for-generator variables into this frame.
       if (slotInfo != null && slotInfo.equals(SymbolTable.FOR_GENERATOR_VARIABLE)) {
-        builder.addSlot(
-            hostFrameDesecriptor.getSlotKind(i), hostFrameDesecriptor.getSlotName(i), null);
+        hasForGenVars = true;
+        forGenVars.add(
+            new Slot(hostFrameDesecriptor.getSlotKind(i), hostFrameDesecriptor.getSlotName(i)));
+      } else {
+        forGenVars.add(new Slot(hostFrameDesecriptor.getSlotKind(i), Identifier.DUMMY));
+      }
+    }
+    // only add for generator vars if there's any
+    if (hasForGenVars) {
+      for (var slot : forGenVars) {
+        builder.addSlot(slot.kind, slot.name, null);
```

Comment:
We can avoid the extra allocation here; can just start filling things in the first time we encounter a for generator variable.

```suggestion
        if (!hasForGenVars) {
          hasForGenVars = true;
          for (var j = 0; j < i; j++) {
            builder.addSlot(FrameSlotKind.Illegal, new Object(), null);
          }
        }
        builder.addSlot(hostFrameDesecriptor.getSlotKind(i), hostFrameDesecriptor.getSlotName(i), null);
      } else if (hasForGenVars) {
        builder.addSlot(FrameSlotKind.Illegal, new Object(), null);
```

Reply:
Good one.